File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/dialog/AddIngredientDialog.java
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.util.Constants;
import de.karzek.diettracker.presentation.util.StringUtils;
import de.karzek.diettracker.presentation.util.ValidationUtil;

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
public class AddIngredientDialog extends AppCompatDialogFragment {

    private static final float LOWER_BOUND_AMOUNT = 0.0f;
    private static final float UPPER_BOUND_AMOUNT = 2000.0f;

    private static final String EXTRA_MANUAL_INGREDIENT_ID = "EXTRA_MANUAL_INGREDIENT_ID";
    private static final String EXTRA_UNITS = "EXTRA_UNITS";
    public static final String EXTRA_GROCERY_QUERY = "EXTRA_GROCERY_QUERY";
    public static final String EXTRA_AMOUNT = "EXTRA_AMOUNT";
    public static final String EXTRA_UNIT_ID = "EXTRA_UNIT_ID";

    @BindView(R.id.ingredient_amount)
    EditText amount;
    @BindView(R.id.spinner_unit)
    Spinner spinner;
    @BindView(R.id.ingredient_product_name)
    EditText groceryQuery;

    @BindView(R.id.dialog_action_dismiss)
    Button dismiss;
    @BindView(R.id.dialog_action_add_ingredient)
    Button addIngredient;

    private View view;
    private ArrayList<String> units;
    private int manualIngredientId;

    private OnAddIngredientClickedInDialogListener addListener;
    private OnSaveIngredientClickedInDialogListener saveListener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.dialog_add_ingredient, container, false);

        ButterKnife.bind(this, view);

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            units = bundle.getStringArrayList(EXTRA_UNITS);
            manualIngredientId = bundle.getInt(EXTRA_MANUAL_INGREDIENT_ID);
            initializeSpinner();
        }

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                addListener = null;
                saveListener = null;
            }
        });

        if (manualIngredientId == Constants.INVALID_ENTITY_ID)
            addIngredient.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (inputFieldsValid()) {
                        int selectedUnitId = spinner.getSelectedItemPosition();
                        dismiss();
                        if (amount.getText().toString().equals(""))
                            addListener.onAddManualIngredientClicked(Float.valueOf(amount.getHint().toString()), selectedUnitId, groceryQuery.getText().toString());
                        else
                            addListener.onAddManualIngredientClicked(Float.valueOf(amount.getText().toString()), selectedUnitId, groceryQuery.getText().toString());
                        addListener = null;
                    }
                }
            });
        else {
            addIngredient.setText(R.string.save_button);
            addIngredient.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (inputFieldsValid()) {
                        int selectedUnitId = spinner.getSelectedItemPosition();
                        dismiss();
                        if (amount.getText().toString().equals(""))
                            saveListener.onSaveManualIngredientClicked(manualIngredientId, Float.valueOf(amount.getHint().toString()), selectedUnitId, groceryQuery.getText().toString());
                        else
                            saveListener.onSaveManualIngredientClicked(manualIngredientId, Float.valueOf(amount.getText().toString()), selectedUnitId, groceryQuery.getText().toString());
                        saveListener = null;
                    }
                }
            });

            groceryQuery.setText(bundle.getString(EXTRA_GROCERY_QUERY));
            amount.setText(StringUtils.formatFloat(bundle.getFloat(EXTRA_AMOUNT)));
            spinner.setSelection(bundle.getInt(EXTRA_UNIT_ID));
        }

        return view;
    }

    private boolean inputFieldsValid() {
        if (groceryQuery.getText().toString().equals("")) {
            showInvalidFieldsError();
            return false;
        } else {
            if(!StringUtils.isNullOrEmpty(amount.getText().toString())) {
                if (ValidationUtil.isValid(LOWER_BOUND_AMOUNT, UPPER_BOUND_AMOUNT, Float.valueOf(amount.getText().toString()), amount, getContext()))
                    return true;
                else
                    return false;
            } else {
                return true;
            }
        }
    }

    private void showInvalidFieldsError() {
        groceryQuery.setError(getString(R.string.error_message_missing_product_name));
    }

    private void initializeSpinner() {
        ArrayAdapter<String> unitAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item, units);
        unitAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(unitAdapter);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            addListener = (OnAddIngredientClickedInDialogListener) getActivity();
            saveListener = (OnSaveIngredientClickedInDialogListener) getActivity();
        } catch (ClassCastException e) {
            throw new ClassCastException("fragment must implement OnAddIngredientClickedInDialogListener and OnSaveIngredientClickedInDialogListener");
        }
    }

    public interface OnAddIngredientClickedInDialogListener {
        void onAddManualIngredientClicked(float amount, int selectedUnitId, String groceryQuery);
    }

    public interface OnSaveIngredientClickedInDialogListener {
        void onSaveManualIngredientClicked(int id, float amount, int selectedUnitId, String groceryQuery);
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/recipeManipulation/dialog/AddIngredientDialog.kt
package de.karzek.diettracker.presentation.main.cookbook.recipeManipulation.dialog

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatDialogFragment
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.util.Constants
import de.karzek.diettracker.presentation.util.StringUtils
import de.karzek.diettracker.presentation.util.ValidationUtil
import kotlinx.android.synthetic.main.dialog_add_ingredient.*
import java.util.*

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
class AddIngredientDialog : AppCompatDialogFragment() {

    private var dialogView: View? = null
    private var units: ArrayList<String>? = null
    private var manualIngredientId: Int = 0

    private var addListener: OnAddIngredientClickedInDialogListener? = null
    private var saveListener: OnSaveIngredientClickedInDialogListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialogView = inflater.inflate(R.layout.dialog_add_ingredient, container, false)

        val bundle = this.arguments
        if (bundle != null) {
            units = bundle.getStringArrayList(EXTRA_UNITS)
            manualIngredientId = bundle.getInt(EXTRA_MANUAL_INGREDIENT_ID)
            initializeSpinner()
        }

        dialog_action_dismiss.setOnClickListener {
            dismiss()
            addListener = null
            saveListener = null
        }

        if (manualIngredientId == Constants.INVALID_ENTITY_ID)
            dialog_action_add_ingredient.setOnClickListener {
                if (inputFieldsValid()) {
                    val selectedUnitId = spinner_unit.selectedItemPosition
                    dismiss()
                    if (ingredient_amount.text.toString() == "")
                        addListener!!.onAddManualIngredientClicked(java.lang.Float.valueOf(ingredient_amount.hint.toString()), selectedUnitId, ingredient_product_name.text.toString())
                    else
                        addListener!!.onAddManualIngredientClicked(java.lang.Float.valueOf(ingredient_amount.text.toString()), selectedUnitId, ingredient_product_name.text.toString())
                    addListener = null
                }
            }
        else {
            dialog_action_add_ingredient.setText(R.string.save_button)
            dialog_action_add_ingredient.setOnClickListener {
                if (inputFieldsValid()) {
                    val selectedUnitId = spinner_unit.selectedItemPosition
                    dismiss()
                    if (ingredient_amount.text.toString() == "")
                        saveListener!!.onSaveManualIngredientClicked(manualIngredientId, java.lang.Float.valueOf(ingredient_amount.hint.toString()), selectedUnitId, ingredient_product_name.text.toString())
                    else
                        saveListener!!.onSaveManualIngredientClicked(manualIngredientId, java.lang.Float.valueOf(ingredient_amount.text.toString()), selectedUnitId, ingredient_product_name.text.toString())
                    saveListener = null
                }
            }

            ingredient_product_name.setText(bundle!!.getString(EXTRA_GROCERY_QUERY))
            ingredient_amount.setText(StringUtils.formatFloat(bundle.getFloat(EXTRA_AMOUNT)))
            spinner_unit.setSelection(bundle.getInt(EXTRA_UNIT_ID))
        }

        return dialogView
    }

    private fun inputFieldsValid(): Boolean {
        return if (ingredient_product_name.text.toString() == "") {
            showInvalidFieldsError()
            false
        } else {
            if (!StringUtils.isNullOrEmpty(ingredient_amount.text.toString())) {
                ValidationUtil.isValid(LOWER_BOUND_AMOUNT, UPPER_BOUND_AMOUNT, java.lang.Float.valueOf(ingredient_amount.text.toString()), ingredient_amount, context)
            } else {
                true
            }
        }
    }

    private fun showInvalidFieldsError() {
        ingredient_product_name.error = getString(R.string.error_message_missing_product_name)
    }

    private fun initializeSpinner() {
        val unitAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, units!!)
        unitAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner_unit.adapter = unitAdapter
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            addListener = activity as OnAddIngredientClickedInDialogListener?
            saveListener = activity as OnSaveIngredientClickedInDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("fragment must implement OnAddIngredientClickedInDialogListener and OnSaveIngredientClickedInDialogListener")
        }

    }

    interface OnAddIngredientClickedInDialogListener {
        fun onAddManualIngredientClicked(amount: Float, selectedUnitId: Int, groceryQuery: String)
    }

    interface OnSaveIngredientClickedInDialogListener {
        fun onSaveManualIngredientClicked(id: Int, amount: Float, selectedUnitId: Int, groceryQuery: String)
    }

    companion object {

        private const val LOWER_BOUND_AMOUNT = 0.0f
        private const val UPPER_BOUND_AMOUNT = 2000.0f

        private const val EXTRA_MANUAL_INGREDIENT_ID = "EXTRA_MANUAL_INGREDIENT_ID"
        private const val EXTRA_UNITS = "EXTRA_UNITS"
        const val EXTRA_GROCERY_QUERY = "EXTRA_GROCERY_QUERY"
        const val EXTRA_AMOUNT = "EXTRA_AMOUNT"
        const val EXTRA_UNIT_ID = "EXTRA_UNIT_ID"
    }
}
